### DISCUSSION TITLE
write log messages to stderr

### DISCUSSION THREAD

**Discussion by finswimmer** [sqlalchemy/sqlalchemy#5878] [Usage Questions] [answered]
Hello everyone.

**Describe the bug**
At the moment `sqlalchemy` prints out messages to `stdout` by default, e.g when setting `echo=True` in `create_engine`. This is also stated out clearly in [docs](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine).

Messages by `sqlalchemy` are usually useful for diagnostics. According to https://www.gnu.org/software/libc/manual/html_node/Standard-Streams.html it is recommended to use `stderr` for such a purpose. This becomes more important, if the program, in which `sqlalchemy` is used, writes its output to `stdout`, as there is no (easy) way which messages are a result of the program and which one are debug messages by `sqlalchemy`.

**Expected behavior**
Messages by `sqlalchemy` should be printed out to `stderr` by default, or at least this should be configurable.

Is [this](https://github.com/sqlalchemy/sqlalchemy/blob/3df145c57c966e1511bccc117161563123ec6f0b/lib/sqlalchemy/log.py#L34) the correct line that is responsable for the current behavior?

fin swimmer

**Comment by zzzeek** [comment, marked as answer]
hey there -

echo=True is not intended as a production-level logging facility, it's used for when you're in a Python console and want to see SQL generated quickly.

For production-quality logging, you should use Python's standard [logging](https://docs.python.org/3/library/logging.html) system.  Configuration of SQLAlchemy's logging in terms of this system is described at: https://docs.sqlalchemy.org/en/13/core/engines.html#configuring-logging

**Reply by rbanffy** [reply]
It'd still nice to have that option when you are piping output of a command-line to another tool. Sometimes you leave echo on for debugging, but would prefer it not to interfere with stdout. In this case, changing the API slightly to accept a stream object instead of a boolean would do the trick nicely.
